br/pkg/aws/ebs.go

```diff
@@ -0,0 +1,236 @@
+	pendingSnapshots := make([]*string, 0, len(snapIDMap))
+	for volID := range snapIDMap {
+		snapID := snapIDMap[volID]
+		pendingSnapshots = append(pendingSnapshots, &snapID)
+	}
+	totalVolumeSize := int64(0)
+
+	log.Info("starts check pending snapshots", zap.Any("snapshots", pendingSnapshots))
+	for {
+		if len(pendingSnapshots) == 0 {
+			log.Info("all pending volume snapshots are finished.")
+			return totalVolumeSize, nil
+		}
+
+		// check pending snapshots every 5 seconds
+		time.Sleep(5 * time.Second)
+		log.Info("check pending snapshots", zap.Int("count", len(pendingSnapshots)))
+		resp, err := e.ec2.DescribeSnapshots(&ec2.DescribeSnapshotsInput{
+			SnapshotIds: pendingSnapshots,
+		})
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		var uncompletedSnapshots []*string
+		for _, s := range resp.Snapshots {
+			if *s.State == ec2.SnapshotStateCompleted {
+				log.Info("snapshot completed", zap.String("id", *s.SnapshotId))
+				totalVolumeSize += *s.VolumeSize
+				progress.Inc()
+			} else {
+				log.Debug("snapshot creating...", zap.Stringer("snap", s))
+				uncompletedSnapshots = append(uncompletedSnapshots, s.SnapshotId)
+			}
+		}
+		pendingSnapshots = uncompletedSnapshots
+	}
+}
+
+func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
+	pendingSnaps := make([]*string, 0, len(snapIDMap))
+	for volID := range snapIDMap {
+		snapID := snapIDMap[volID]
+		pendingSnaps = append(pendingSnaps, &snapID)
+	}
+
+	for _, snapID := range pendingSnaps {
+		// todo: concurrent delete
+		_, err2 := e.ec2.DeleteSnapshot(&ec2.DeleteSnapshotInput{
+			SnapshotId: snapID,
+		})
+		if err2 != nil {
+			log.Error("failed to delete volume", zap.Error(err2))
+			// todo: we can only retry for a few times, might fail still, need to handle error from outside.
+		}
+	}
+}
+
+// CreateVolumes create volumes from snapshots
+// if err happens in the middle, return half-done result
+// returned map: store id -> old volume id -> new volume id
+func (e *EC2Session) CreateVolumes(meta *config.EBSBasedBRMeta, volumeType string, iops, throughput int64) (map[string]string, error) {
+	template := ec2.CreateVolumeInput{VolumeType: &volumeType}
+	if iops > 0 {
+		template.SetIops(iops)
+	}
+	if throughput > 0 {
+		template.SetThroughput(throughput)
+	}
+
+	newVolumeIDMap := make(map[string]string)
+	for _, store := range meta.TiKVComponent.Stores {
+		for _, oldVol := range store.Volumes {
+			// TODO: build concurrent requests here.
+			log.Debug("create volume from snapshot", zap.Any("volume", oldVol))
+			req := template
+			req.SetSnapshotId(oldVol.SnapshotID)
+			resp, err := e.ec2.CreateVolume(&req)
+			if err != nil {
+				return newVolumeIDMap, errors.Trace(err)
+			}
+			log.Info("new volume creating", zap.Stringer("snap", resp))
+			newVolumeIDMap[oldVol.ID] = *resp.VolumeId
+		}
+	}
+	return newVolumeIDMap, nil
+}
+
+func (e *EC2Session) WaitVolumesCreated(volumeIDMap map[string]string, progress glue.Progress) (int64, error) {
```

Comment:
can we unified `WaitVolumesCreated` and `WaitSnapshotFinished` since they have the same framework.

Reply:
[rename](https://github.com/pingcap/tidb/pull/36078/commits/ba5ceb6f65f71402e1a422431e221d6bf9b59d1a)